Src/IronPython/Runtime/WeakRef.cs

```diff
@@ -91,67 +94,133 @@ struct CallbackInfo {
         }
 
         public void RemoveHandlerAt(int index) {
-            _callbacks.RemoveAt(index);
+            _lock.EnterWriteLock();
+            try
+            {
+                _callbacks.RemoveAt(index);
+            }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
         }
 
         public void RemoveHandler(object o) {
-            for (int i = 0; i < HandlerCount; i++) {
-                if (GetWeakRef(i) == o) {
-                    RemoveHandlerAt(i);
-                    break;
+            _lock.EnterWriteLock();
+            try
+            {
+                for (int i = 0; i < _callbacks.Count; i++) {
+                    if (_callbacks[i].WeakRef == o)
+                    {
+                        _callbacks.RemoveAt(i);
+                        break;
+                    }
                 }
             }
+            finally
+            {
+                _lock.ExitWriteLock();
+            }
+
         }
 
         public object GetHandlerCallback(int index) {
-            return _callbacks[index].Callback;
+            _lock.EnterReadLock();
+            try
+            {
+                return _callbacks[index].Callback;
+            }
+            finally
+            {
+                _lock.ExitReadLock();
+            }
         }
 
         public object GetWeakRef(int index) {
-            return _callbacks[index].WeakRef;
+            _lock.EnterReadLock();
+            try
+            {
+                return _callbacks[index].WeakRef;
+            }
+            finally
+            {
+                _lock.ExitReadLock();
+            }
         }
 
         ~WeakRefTracker() {
-            // callbacks are delivered last registered to first registered.
-            for (int i = _callbacks.Count - 1; i >= 0; i--) {
-
-                CallbackInfo ci = _callbacks[i];
-                try {
-                    if (ci.Callback != null) {
-                        // a little ugly - we only run callbacks that aren't a part
-                        // of cyclic trash.  but classes use a single field for
-                        // finalization & GC - and that's always cyclic, so we need to special case it.
-                        InstanceFinalizer fin = ci.Callback as InstanceFinalizer;
-                        if (fin != null) {
-                            // Going through PythonCalls / Rules requires the types be public.
-                            // Explicit check so that we can keep InstanceFinalizer internal.
-                            fin.CallDirect(DefaultContext.Default);
-                                
-                        } else {
-                            // Non-instance finalizer goes through normal call mechanism.
-                            object weakRef = ci.WeakRef;
-                            if (weakRef != null)
-                                PythonCalls.Call(ci.Callback, weakRef);
+            _lock.EnterWriteLock();
```

Comment:
Is there a reason why this one can't be a read lock?

Reply:
I guess its intent rather than actual action. This block is trying run all the callbacks and then remove them as far as I can see - while it doesn't actually edit the list, there's an implicit _callbacks.Clear() since this is a finaliser. My personal style would be to be conservative and assume this code is "entitled" to edit the list and therefore protect it as if that were the case (and possibly prevent a future bug if someone does edit the list) . Happy to change it if you prefer to leave it as a read lock.

It probably doesn't matter either way since the finaliser should be called when there are no references to the WeakRefTracker. Possibly it doesn't need a lock at all...